Propogate dynamo shape_env to make_fx

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96561
* __->__ #96437

Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Differential Revision: [D44561753](https://our.internmc.facebook.com/intern/diff/D44561753)

tugsbayasgalan committed:
Propogate dynamo shape_env to make_fx

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -698,7 +705,11 @@ def result_capturing_wrapper(*graph_inputs):
         opt_f = optimize_assert(
             dynamo_normalization_capturing_compiler,
-            hooks=Hooks(guard_export_fn=guard_export_print, guard_fail_fn=None),
+            hooks=Hooks(
+                guard_export_fn=guard_export_print,
+                guard_fail_fn=None,
+                expose_shape_env_fn=expose_shape_env_fn,
+            ),
```
You don't need this hook, I think you can just get it off the tracing context?

tugsbayasgalan replied:
Hmm tracing context is inside OutputGraph(fx.Tracer) which is not available in eval_frame.py where we only see the graph. I think I would have to somehow add the env as attribute in graph to access in export. So i thought this was cleaner.

voznesenskym replied:
It is globally available! Try torch._guards.TracingContext.get()?

tugsbayasgalan replied:
ah gotcha thanks for the pointer

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -670,6 +672,11 @@ def guard_export_print(guards: Set[_guards.Guard]):
+    def expose_shape_env_fn(shape_env: ShapeEnv):
+        nonlocal out_shape_env
+        assert out_shape_env is None, "whole graph should only have one shape env"
+        out_shape_env = shape_env
```
Instead of this, when you are called in dynamo_normalization_capturing_compiler your example inputs will be fake tensors, whose fake tensor mode will have shape env on it. Extract the shape env that way. This is how AOTAutograd does it, by the way.

tugsbayasgalan replied:
I see. that sounds easier

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -657,7 +657,7 @@ def disable_autocast_cache():
 
 
-def make_fx(f, decomposition_table=None, tracing_mode="real", _allow_non_fake_inputs=False):
+def make_fx(f, decomposition_table=None, tracing_mode="real", _allow_non_fake_inputs=False, _shape_env=None):
```
kwarg only

ezyang commented:
This seems incomplete. Because you are reusing the ShapeEnv from Dynamo, you need to ALSO reuse the fake tensors from dynamo. Which you should have. But then you don't need to do tracing_mode="symbolic"; you can pass in the fake tensors directly, and use tracing_mode="real", and make_fx will implicitly use the ambient fake mode from them. Basically, you want to do the same thing that AOTAutograd does.

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the ShapeEnv dynamo used to make_fx. Next step would be unifying all different configs in a central place. Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx. Therefore, the new make_fx behaviour would be:
1. If example inputs are passed in, make_fx will use the FakeTensorMode from those inputs for all tracing modes.
2. If example inputs are not passed in, make_fx will decide what mode to use using its' old logic. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43994693).

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx. Therefore, the new make_fx behaviour would be:
1. If example inputs are passed in, make_fx will use the FakeTensorMode from those inputs for all tracing modes.
2. If example inputs are not passed in, make_fx will decide what mode to use using its' old logic. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43994693).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx. Therefore, the new make_fx behaviour would be:
1. If example inputs are passed in, make_fx will use the FakeTensorMode from those inputs for all tracing modes.
2. If example inputs are not passed in, make_fx will decide what mode to use using its' old logic. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

jansel commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -657,31 +657,50 @@ def disable_autocast_cache():
 
 
-def make_fx(f, decomposition_table=None, tracing_mode="real", _allow_non_fake_inputs=False):
+def make_fx(f, decomposition_table=None, tracing_mode="real", *, _allow_non_fake_inputs=False, _example_inputs=None):
```
Adding `_example_inputs` would seem to create a kind of awkward usage:
```
gm = make_fx(fn, _example_inputs=[a,b,c])(a,b,c)
``` 
Does make_fx even need the non-fake inputs?

ezyang replied:
Agreed, although I see the technical problem that resulted in this. Thinking...

ezyang replied:
OK, so, if you do this exactly the way AOTAutograd does it, you don't need to take in `example_inputs` in `make_fx`, because you are going to call it with `tracing_mode="real"`, and the calling code is going to take care of all the ShapeEnv creation, fake tensor mode entering, and input translation.

So perhaps the best way to do this, is to make a wrapper around `make_fx`, that calls `make_fx` in the same way, but does all this stuff. The reason we can't make `make_fx` do this is because it's staged so that you don't take the inputs immediately, and this is incompatible with any API style where we work things out from the inputs.

ezyang replied:
Actually it's not even clear to me how much of a wrapper you even need, given that you should already have fake tensors ambiently available. Just call `tracing_mode="real"`! Do you even need this PR? Try the E2E.

tugsbayasgalan replied:
Yeah I think i got it working with just real mode.

ezyang commented on torch/overrides.py:
```diff
@@ -185,6 +185,7 @@ def get_ignored_functions() -> Set[Callable]:
         torch.sym_max,
         torch.sym_min,
         torch.sym_not,
+        torch.is_concrete_int,
```
I would NOT expose this from top-level torch.* namespace

tugsbayasgalan replied:
Sounds good

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx. Therefore, the new make_fx behaviour would be:
1. If example inputs are passed in, make_fx will use the FakeTensorMode from those inputs for all tracing modes.
2. If example inputs are not passed in, make_fx will decide what mode to use using its' old logic. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43994693).

ezyang commented on functorch/experimental/_cond.py:
```diff
@@ -40,7 +41,7 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
     assert all(isinstance(o, torch.Tensor) for o in operands), "Cond operands must be a list of tensors"
 
-    with disable_proxy_modes_tracing():
+    with disable_proxy_modes_tracing(), maybe_disable_all_fake_tensor_modes():
```
Can you explain why you have to disable fake mode here

ezyang reviewed (changes requested):
need to discuss some things here

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -705,10 +733,26 @@ def wrap_fake(x):
+                if _allow_non_fake_inputs:
+                    new_fake_mode = FakeTensorMode(
+                        allow_fallback_kernels=x.fake_mode.allow_fallback_kernels,
+                        allow_non_fake_inputs=True,
```
This seems highly questionable. The original allocation of FakeTensorMode should have gotten this setup; fixing it up after the fact is something is not supported API.

tugsbayasgalan replied:
Yeah i just moved it up to dynamo, because only reason it exist is to workaround params and buffers. In export, we don't care about their mutation for now (until Brian's refactor), so i think it is ok to just set it to true in export mode only, and remove it later.

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43994693).

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -113,6 +113,20 @@ def has_hint(a):
+    """
+    assert isinstance(a, SymInt) or isinstance(a, int)
+    if isinstance(a, SymInt):
+        return a.node.is_int()
```
This is wrong, a SymInt will always return True here.

tugsbayasgalan replied:
What I am trying to test here is if the backing value for SymInt is a concrete integer, instead of a symbol.

ezyang replied:
yes, I think what I'd do is peak at the sympy expr and see if it's Integer

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -761,12 +768,17 @@ def graph_with_interpreter(*args):
-        )(*graph_captured_input)
+        fake_tensor_mode = null_context()
+        if example_fake_inputs is not None and len(example_fake_inputs) > 0:
+            fake_tensor_mode = example_fake_inputs[0].fake_mode
```
nit: it is better if you follow the aot autograd pattern exactly, which is to iterate over the fake inputs and accept the first one that actually has a fake mode. This handles the case when the first input is not actually a tensor (if that could occur)

ezyang commented on functorch/experimental/_cond.py:
```diff
@@ -164,7 +164,8 @@ def _has_potential_branch_input_mutation(branch, inputs):
     """
     try:
-        gm = make_fx(branch)(*inputs)
+        with disable_proxy_modes_tracing():
```
Tell me more about this change?

tugsbayasgalan replied:
not needed anymore

ezyang commented:
This looks much better, high level code looks good, some substantive comments though.

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43994693).

ezyang commented on functorch/experimental/_cond.py:
```diff
@@ -191,6 +191,7 @@ def _has_potential_branch_input_alias(branch, inputs):
     """
     try:
         gm = make_fx(branch)(*inputs)
+
```
nit: avoid random whitespace changes like this

ezyang commented on torch/ao/quantization/_pt2e/utils.py:
```diff
@@ -17,7 +17,7 @@ def _get_tensor_constant_from_node(node, m):
 def _fuse_conv_bn_(m: GraphModule) -> None:
     for n in m.graph.nodes:
-        if n.op != "call_function" or n.target != torch.ops.aten.native_batch_norm.default:
+        if n.op != "call_function" or n.target != torch.ops.aten._native_batch_norm_legit_no_training.default:
```
blegh, you sure about this?

ezyang commented on test/dynamo/test_export.py:
```diff
@@ -2275,6 +2276,28 @@ def forward(self, *args):
+                    self.assertTrue(is_concrete_int(shape))
+
+        # this should be captured as static, as export won't generate any symbols.
+        self.assertEqual(gm(torch.ones(2, 4)), torch.ones(2, 4).sin())
```
It doesn't seem like this tests the propagation of fake mode though?

ezyang reviewed (approved):
okey dokey

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43994693).

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -f "Landed in fbcode already"

clee2000 commented:
@pytorchbot revert -m "broke dynamic shapes tests https://hud.pytorch.org/pytorch/pytorch/commit/3a22916c7a501499eec9053e3a568f2b1f49938c https://github.com/pytorch/pytorch/actions/runs/4558606911/jobs/8041864639, I guess this is technically ignored signal?" -c ignoredsignal

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D43994693](https://our.internmc.facebook.com/intern/diff/D43994693)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire


[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44561753).

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D44561753](https://our.internmc.facebook.com/intern/diff/D44561753)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Propogate dynamo shape_env to make_fx"


Currently, when we use assume_static_by_default flag, dynamo won't produce any symbols for input tensors. But when we pass the dynamo generated graph onto make_fx via torchdynamo.export(aten_graph=True), there is no way to pass this flag. We enable this by directly passing the fake tensors dynamo used to make_fx and call make_fx with "real" mode with fake tensors from dynamo. 

Note that this is modified version of (https://github.com/pytorch/pytorch/pull/96143)


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

Differential Revision: [D44561753](https://our.internmc.facebook.com/intern/diff/D44561753)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44561753).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan commented:
@pytorchbot merge -f "Landed fbcode"

bigfootjon commented:
@pytorchbot merge -f "Landed fbcode"
